File: akk_example/src/main/scala/com/hdu/supervisor/SupervisorActor.scala
package com.hdu.supervisor

import akka.actor.SupervisorStrategy.{Escalate, Restart, Resume, Stop}
import akka.actor.{Actor, OneForOneStrategy, Props}

/**
  * DESCRIPTION:
  *
  * @author shushoufu
  * @Date 2018/4/10
  * @Time 下午8:42
  */
class SupervisorActor extends Actor{
  import scala.concurrent.duration._


  override val supervisorStrategy =
    OneForOneStrategy(maxNrOfRetries = 10,withinTimeRange = 1 minutes){
      case _:ArithmeticException => Resume
      case _:NullPointerException => Restart
      case _:IllegalArgumentException => Stop
      case _:Exception => Escalate
    }

  override def receive: Receive = {
    case p:Props => sender() ! context.actorOf(p)
  }
}

File: akk_example/src/main/java/com/hdu/chapter3_supervisorStrategy/SupervisorActor.java
package com.hdu.chapter3_supervisorStrategy;

import akka.actor.*;
import akka.japi.pf.DeciderBuilder;
import scala.concurrent.duration.Duration;

import java.util.concurrent.TimeUnit;

/**
 * DESCRIPTION:
 *
 * @author shushoufu
 * @Date 2018/6/20
 * @Time 上午10:52
 */
public class SupervisorActor extends AbstractActor {

    public static Props props() {
        return Props.create(SupervisorActor.class);
    }

    private SupervisorStrategy strategy = new OneForOneStrategy(10, Duration.create(1, TimeUnit.MINUTES),
            DeciderBuilder
                    .match(ArithmeticException.class, e -> SupervisorStrategy.resume())
                    .match(NullPointerException.class, e -> SupervisorStrategy.restart())
                    .match(MyException.class, e -> {
                        System.out.println("the actor " + getSender().path() + " get a exception" + e.getMessage());
                        return SupervisorStrategy.stop();
                    })
                    .match(IllegalArgumentException.class, e -> SupervisorStrategy.stop())
                    .matchAny(o -> SupervisorStrategy.escalate())
                    .build());
//    private SupervisorStrategy strategy = new AllForOneStrategy(10, Duration.create(1, TimeUnit.MINUTES),
//            DeciderBuilder
//                    .match(ArithmeticException.class,e-> SupervisorStrategy.resume())
//                    .match(NullPointerException.class,e->SupervisorStrategy.restart())
//                    .match(IllegalArgumentException.class,e->SupervisorStrategy.stop())
//                    .matchAny(o->SupervisorStrategy.escalate())
//                    .build());

    @Override
    public SupervisorStrategy supervisorStrategy() {
        return strategy;
    }

    @Override
    public Receive createReceive() {
        return receiveBuilder()
                .match(Props.class, props -> {
                    System.out.println("send is " + getSender().path());
                    ActorRef childActor = getContext().actorOf(props, "child");
                    getContext().watch(childActor);
                    getSender().tell(childActor, getSelf());

                })
                .match(Terminated.class, r -> {
                    System.out.println("the watched actor stopped, actor information is " + r.getActor());
                })
                .build();
    }
}
